### ISSUE TITLE
Change default value of allowTernary for no-unused-expressions

### ISSUE THREAD

**Issue by HRK44** [eslint/eslint#10625]
**What rule do you want to change?**
I want to change ``no-unused-expressions``, and set the default value of ``allowTernary`` to ``true``. I don't see how an expression is unused in a ternary expression or at least, I would argue that this uncommon.

**Does this change cause the rule to produce more or fewer warnings?**
This should produce less warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**
New default behavior

**Please provide some example code that this change will affect:**

```js
let isThisTrue = funcThatReturnsABoolean();
isThisTrue ? doThis() : doThat();
```

**What does the rule currently do for this code?**
It will generate an error/warning about ``no-unused-expressions`` for ``isThisTrue``.

**What will the rule do after it's changed?**
It won't add a warning.

**Comment by HRK44** [comment]
Can @tjbenton explain the thumbdown? Has anyone reviewed this issue?

Thanks

**Comment by platinumazure** [comment]
Hi @HRK44, thanks for the issue and apologies for letting this slip through the cracks.

I can't speak for @tjbenton. As a maintainer, I'm mildly opposed to this change because it is a breaking change that, as far as I can tell, is done mostly for convenience. I am open to being persuaded otherwise, and if the rest of ESLint team is okay with this proposal I will defer to consensus. However, every breaking change is costly to our users and we have *a lot* of users, so I don't want to make a breaking change without a very good reason.

**Comment by HRK44** [comment]
@platinumazure Thanks for reviewing.

To add more background on why I'm suggesting this, if you look at the code snippet I have put, for me it was very surprising that I got an ``no-unused-expressions`` on ``isThisTrue`` since I literally just used it one line after - then I stumbled upon the ``allowTernary : false`` option.

Could someone explain the reason of the existence of ``allowTernary`` and/or an use-case of the ``allowTernary: false`` over ``allowTernary: true`` where the variable is ``unused`` ? That would be helpful.

Thanks

**Comment by ljharb** [comment]
I don’t think it should change.

Code in a ternary branch might be used, or might not - the current default is the safer one, which is that it’s not considered used if it’s 100% certainly used.

**Comment by HRK44** [comment]
@ljharb In my code snippet, ``isThisTrue`` is 100% used since it is used to determine which action to call.

**Comment by ljharb** [comment]
@HRK44 what i mean is, it's impossible to statically determine whether `isThisTrue` is true, false, always true, or always false - so it's impossible to safely determine if those variables are used or not.

**Comment by not-an-aardvark** [comment]
@ljharb I'm not sure I understand your point -- the same would apply when using a statement like this:

```js
if (isThisTrue) {
    doThis();
} {
    doThat();
}
```

You're correct that we can't tell whether both branches are reachable, but I don't see how that's related to the `no-unused-expressions` rule, because the `no-unused-expressions` rule isn't really intended to detect unreachable code.

**Comment by ljharb** [comment]
Ah, you're right, I'm thinking of `no-unused-variables`. I'll hide my comments as off-topic.

**Comment by tjbenton** [comment]
@HRK44 To me changing the default to allow this will promote bad practices and reduce legibility in our code. A `if`/`else` statement is much easier to read as a developer which makes it easier to maintain overtime and over a large codebase. I think the ternary operator should only be used to assign values to variables or in return statements. Like all the examples on [mozilla](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator) they are using the ternary operator to conditionally assign variables or in return statements.

**Comment by HRK44** [comment]
Thanks for your input. After reading the specs, it seems that the ternary operator should return something - thus triggering the ``no-unused-expression`` in my example. I guess this works fine then with the current state. Feel free to re-open if you think this is still worth evaluating.
